packages/apollo-server-hapi/src/hapiApollo.ts

```diff
@@ -34,56 +39,100 @@ const graphqlHapi: IPlugin = {
+      try {
+        let query: HttpQueryRequest;
 
-          const err = new Boom(error.message, { statusCode: error.statusCode });
-          if (error.headers) {
-            Object.keys(error.headers).forEach(header => {
-              err.output.headers[header] = error.headers[header];
-            });
-          }
-          // Boom hides the error when status code is 500
-          err.output.payload.message = error.message;
-          throw err;
+        if (request.method === 'get') {
+          query = request.query as any;
+        } else if (isUploadEnabled && request.mime === 'multipart/form-data') {
+          query = (await processFileUploads(
+            request.raw.req,
+            request.raw.res,
+            options.uploadsConfig,
+          )) as any;
+        } else {
+          // The request is unparsed due to the 'POST' route's config. Use
+          // Subtext, hapi's default parser, to parse the request.
+          // https://github.com/hapijs/subtext
+          const { payload } = await Subtext.parse(request.raw.req, null, {
+            parse: true,
+            output: 'data',
+          });
+          query = payload;
         }
+
+        const { graphqlResponse, responseInit } = await runHttpQuery(
+          [request, h],
+          {
+            method: request.method.toUpperCase(),
+            options: options.graphqlOptions,
+            query,
+            request: convertNodeHttpToRequest(request.raw.req),
+          },
+        );
+
+        const response = h.response(graphqlResponse);
+        Object.keys(responseInit.headers).forEach(key =>
+          response.header(key, responseInit.headers[key]),
+        );
+        return response;
+      } catch (error) {
+        if ('HttpQueryError' !== error.name) {
+          throw Boom.boomify(error);
+        }
+
+        if (true === error.isGraphQLError) {
+          const response = h.response(error.message);
+          response.code(error.statusCode);
+          response.type('application/json');
+          return response;
+        }
+
+        const err = new Boom(error.message, { statusCode: error.statusCode });
+        if (error.headers) {
+          Object.keys(error.headers).forEach(header => {
+            err.output.headers[header] = error.headers[header];
+          });
+        }
+        // Boom hides the error when status code is 500
+        err.output.payload.message = error.message;
+        throw err;
+      }
+    };
+
+    // hapi doesn't allow `options.payload` configuration to non-POST routes.
+    // Register two routes to apply the necessary `options.payload`
+    // configuration to the POST handler.
+    server.route([
+      {
+        method: 'GET',
+        path,
+        vhost: options.vhost || undefined,
+        options: options.route || {},
+        handler,
+      },
+      {
+        method: 'POST',
+        path,
+        vhost: options.vhost || undefined,
+        handler,
+        options: {
```

Comment:
Here's where it gets a little tricky: This patch necessarily adjusts the `options.route` behavior for the POST route (hapi throws when the `options.payload` is applied to a GET route). This _could_ be considered a breaking change.

What this does: prevents hapi from parsing the `http.IncomingMessage`, and passes the raw stream to the route handler as a readable stream.

See:

* `options.payload.output`: https://hapi.dev/api/?v=18.3.2#route.options.payload.output
* `options.payload.parse`: https://hapi.dev/api/?v=18.3.2#route.options.payload.parse

This means the route handler must manually parse the payload (thus the use of subtext).

Reply:
Could you help me understand the breaking change aspect of this a bit more?

Reply:
Trying to recall what I did here 😅

This package's [`ApolloServer#applyMiddleware`](https://github.com/apollographql/apollo-server/blob/0926b20f49b23b63c438ffbd14b2e5f5f02805cc/packages/apollo-server-hapi/src/ApolloServer.ts#L55-L62) has an [optional `route`](https://github.com/apollographql/apollo-server/blob/0926b20f49b23b63c438ffbd14b2e5f5f02805cc/packages/apollo-server-hapi/src/ApolloServer.ts#L152) configuration that's [passed directly to the hapi plugin if it's defined](https://github.com/apollographql/apollo-server/blob/0926b20f49b23b63c438ffbd14b2e5f5f02805cc/packages/apollo-server-hapi/src/ApolloServer.ts#L135-L140). This PR introduces a change in how the hapi plugin's route handler accepts uploads by changing the route's options.

Before:

https://github.com/apollographql/apollo-server/blob/0926b20f49b23b63c438ffbd14b2e5f5f02805cc/packages/apollo-server-hapi/src/hapiApollo.ts#L41

After (abbreviated):

```typescript
server.route([
  {
    method: 'GET',
    // ...
    options: options.route || {},
  },
  {
    method: 'POST',
    // ...
    options: {
      ...options.route,
      payload: {
        output: 'stream',
        parse: false,
      },
    }
  }
])
```

Notice how a user's configuration for `payload.options` _isn't passed through to the route_. This is a necessary change: `graphql-upload`'s [`processRequest`](https://github.com/jaydenseric/graphql-upload#function-processrequest) (via [processFileUploads](processFileUploads)) expects an unread stream such that it can handle the parsing. This is the only hapi [`route.options.payload` config](https://hapi.dev/api/?v=18.4.0#route.options.payload) that doesn't read the incoming message (that I found, anway). Unfortunately, allowing a user to override this breaks uploads.

This modifies how apollo-server-hapi passes user config down to the underlying route: should it be marked as a breaking change?

Reply:
I _guess_ we could change it to:

```diff
 {
   options: {
     ...options.route,
     payload: {
+      ...(options.route || {}).payload,
        output: 'stream',
        parse: false,
     }
   }
 }
```

to give users that extra control.

Reply:
_Hmmm,_ it looks like users were never able to set `route.payload`. I made a demo on Glitch to test this out: https://glitch.com/edit/#!/scratch-animantarx?path=server.js:44:4

Even with `route.payload` set to empty object, hapi seems to error:

```shell
 AssertionError [ERR_ASSERTION]: Cannot set payload settings on HEAD or GET request: GET /graphql
    at new AssertionError (internal/assert.js:269:11)
# ...
```

(See that "Logs" button in the bottom right.)

I think the user was never able to pass payload, so this wouldn't be a breaking change at all 🎉

Reply:
I can confirm that I'm getting this error trying to override payload, because `payload` should only be set on POST requests, and not HEAD or GET. But currently if you pass payload through as route options it will set on the GET route of /graphql too. 

So yes it's not a breaking change